ant-cli/src/actions/download.rs

```diff
@@ -7,123 +7,192 @@
-    eyre::{eyre, Context, Result},
-    Section,
-};
-use std::path::PathBuf;
+use std::{path::PathBuf, process, time::SystemTime};
+
+pub async fn download(
+    addr: &str,
+    dest_path: &str,
+    client: &mut Client,
+    json: bool,
+) -> Result<(), color_eyre::Report> {
+    let start_time = SystemTime::now();
+    let event_receiver = client.enable_client_events();
+    let (task_summary_thread, stop_summary_collection) = collect_task_summary(event_receiver);
 
-pub async fn download(addr: &str, dest_path: &str, client: &mut Client) -> Result<()> {
     let public_address = str_to_addr(addr).ok();
     let private_address = crate::user_data::get_local_private_archive_access(addr)
         .inspect_err(|e| error!("Failed to get private archive access: {e}"))
         .ok();
 
-    match (public_address, private_address) {
-        (Some(public_address), _) => download_public(addr, public_address, dest_path, client).await,
-        (_, Some(private_address)) => download_private(addr, private_address, dest_path, client).await,
-        _ => Err(eyre!("Failed to parse data address {addr}"))
-            .with_suggestion(|| "Public addresses look like this: 0037cfa13eae4393841cbc00c3a33cade0f98b8c1f20826e5c51f8269e7b09d7")
-            .with_suggestion(|| "Private addresses look like this: 1358645341480028172")
-            .with_suggestion(|| "Try the `file list` command to get addresses you have access to"),
+    let public;
+    let result = match (public_address, private_address) {
+        (Some(public_address), _) => {
+            public = true;
+            download_public(public_address, dest_path, client).await
+        }
+        (_, Some(private_address)) => {
+            public = false;
+            download_private(private_address, dest_path, client).await
+        }
+        _ => {
+            println!("Public addresses look like this: 0037cfa13eae4393841cbc00c3a33cade0f98b8c1f20826e5c51f8269e7b09d7");
+            println!("Private addresses look like this: 1358645341480028172");
+            println!("Try the `file list` command to get addresses you have access to");
+            process::exit(exit_code::INVALID_INPUT_EXIT_CODE);
+        }
+    };
+
+    // wait for upload to complete
+    if let Err(e) = stop_summary_collection.send(()) {
+        error!("Failed to send upload completed event: {e:?}");
     }
+
+    let task_summary = task_summary_thread.await?;
+    let download_output = FileDownloadOutput {
+        exit_code: result
+            .as_ref()
+            .map_or_else(|(err, _)| exit_code::download_exit_code(err), |_| 0),
+        addr: addr.to_string(),
+        public,
+        task_summary,
+        start_time,
+        end_time: SystemTime::now(),
+        failed_files: result
+            .as_ref()
+            .map_or_else(|(_, failed_files)| failed_files.clone(), |_| vec![]),
+    };
+
+    if json {
+        download_output.print_json();
+    } else {
+        download_output.print();
+    }
```

Comment:
Can't this be part of client config so we don't need to carry around a json bool?